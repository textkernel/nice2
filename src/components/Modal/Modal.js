import * as React from 'react';
import PropTypes from 'prop-types';
import ReactModal from 'react-modal';
import { bem } from '../../utils';
import styles from './Modal.scss';

const { block, elem } = bem('Modal', styles);

export const Modal = (props) => {
    const {
        children,
        isOpen,
        contentLabel,
        onRequestClose,
        className,
        portalClassName,
        overlayClassName,
        ...rest
    } = props;

    const { className: portalClass } = block({ className: portalClassName, ...rest });
    const { className: overlayClass } = elem('overlay', {
        ...props,
        elemClassName: overlayClassName,
    });
    const { className: overlayEnteredClass } = elem('overlay--entered', props);
    const { className: overlayExitedClass } = elem('overlay--exited', props);
    const { className: contentClass } = elem('content', { ...props, elemClassName: className });
    const { className: contentEnteredClass } = elem('content--entered', props);
    const { className: contentExitedClass } = elem('content--exited', props);

    return (
        <ReactModal
            isOpen={isOpen}
            contentLabel={contentLabel}
            onRequestClose={onRequestClose}
            closeTimeoutMS={300}
            portalClassName={portalClass}
            overlayClassName={{
                base: overlayClass,
                afterOpen: overlayEnteredClass,
                beforeClose: overlayExitedClass,
            }}
            className={{
                base: contentClass,
                afterOpen: contentEnteredClass,
                beforeClose: contentExitedClass,
            }}
            {...rest}
        >
            {children}
        </ReactModal>
    );
};

/**
 * Inits ReactModal and sets global params for it.
 * @param {string} selector - css selector for the app element
 */
Modal.setAppElement = (selector) => {
    ReactModal.setAppElement(selector);
};

Modal.displayName = 'Modal';

Modal.propTypes = {
    /** elements to be rendered within the modal */
    children: PropTypes.node.isRequired,
    /** The state of the modal */
    isOpen: PropTypes.bool.isRequired,
    /** A title for the modal that will be used by screenreaders */
    contentLabel: PropTypes.string.isRequired,
    /** A function to be called when the modal is closed */
    onRequestClose: PropTypes.func,
    /** Additional class to be applied to the content part */
    className: PropTypes.string,
    /** Additional class to be applied to the overlay */
    overlayClassName: PropTypes.string,
    /** Additional class to be applied to the portal */
    portalClassName: PropTypes.string,
};

Modal.defaultProps = {
    onRequestClose: null,
    className: '',
    overlayClassName: null,
    portalClassName: null,
};
