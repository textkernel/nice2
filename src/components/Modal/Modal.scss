.Modal {
    &__overlay {
        position: fixed;
        top: 0;
        left: 0;
        right: 0;
        bottom: 0;
        background-color: rgba(0, 0, 0, 0.5);
        opacity: 0;
        transition: opacity var(--transition-duration) ease-in-out;
    }

    &__overlay--entered {
        opacity: 1;
    }

    &__overlay--exited {
        opacity: 0;
    }

    &__content {
        position: relative;
        top: 15vh;
        width: 65vw;
        height: 65vh;
        margin: auto;
        max-width: var(--site-container-max-size);
        border: 1px solid var(--border-color-strong);
        border-radius: var(--border-radius);
        background-color: var(--color-background);
        box-shadow: 0 1px 25px 0 rgba(0, 0, 0, 0.8);
        overflow: auto;
        outline: none;
        transform: translateY(50px);
        transition: transform var(--transition-duration);
    }

    &__content--entered {
        transform: translateY(0);
    }

    &__content--exited {
        transform: translateY(50px);
    }
}
