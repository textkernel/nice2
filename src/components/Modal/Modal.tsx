import * as React from 'react';
import ReactModal from 'react-modal';
import { bem } from '../../utils';
import styles from './Modal.scss';

interface Props extends ReactModal.Props {
    /** elements to be rendered within the modal */
    children: ReactNode;
    /** The state of the modal */
    isOpen: boolean;
    /** A title for the modal that will be used by screenreaders */
    contentLabel: string;
    /** A function to be called when the modal is closed */
    onRequestClose?: (
        event: React.MouseEvent<Element, MouseEvent> | React.KeyboardEvent<Element>
    ) => void;
    /** Additional class to be applied to the content part */
    className?: string;
    /** Additional class to be applied to the overlay */
    overlayClassName?: string;
    /** Additional class to be applied to the portal */
    portalClassName?: string;
}

interface Modal<P> extends React.FunctionComponent<P> {
    setAppElement: (appElement: string | HTMLElement) => void;
}

const { block, elem } = bem('Modal', styles);

export const Modal: Modal<Props> = (props) => {
    const {
        children,
        isOpen,
        contentLabel,
        onRequestClose,
        className,
        portalClassName,
        overlayClassName,
        ...rest
    } = props;

    const { className: portalClass } = block({ className: portalClassName, ...rest });
    const { className: overlayClass } = elem('overlay', {
        ...props,
        elemClassName: overlayClassName,
    });
    const { className: overlayEnteredClass } = elem('overlay--entered', props);
    const { className: overlayExitedClass } = elem('overlay--exited', props);
    const { className: contentClass } = elem('content', { ...props, elemClassName: className });
    const { className: contentEnteredClass } = elem('content--entered', props);
    const { className: contentExitedClass } = elem('content--exited', props);

    return (
        <ReactModal
            isOpen={isOpen}
            contentLabel={contentLabel}
            onRequestClose={onRequestClose}
            closeTimeoutMS={300}
            portalClassName={portalClass}
            overlayClassName={{
                base: overlayClass,
                afterOpen: overlayEnteredClass,
                beforeClose: overlayExitedClass,
            }}
            className={{
                base: contentClass,
                afterOpen: contentEnteredClass,
                beforeClose: contentExitedClass,
            }}
            {...rest}
        >
            {children}
        </ReactModal>
    );
};

/**
 * Inits ReactModal and sets global params for it.
 * @param {string} selector - css selector for the app element
 */
Modal.setAppElement = (selector) => {
    ReactModal.setAppElement(selector);
};

Modal.displayName = 'Modal';
