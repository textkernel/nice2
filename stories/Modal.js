import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, text, withKnobs } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';
import { Modal } from '@textkernel/oneui';

storiesOf('Modal', module)
    .addDecorator(withKnobs)
    .addDecorator(withInfo)
    .add(
        'Modal',
        () => {
            const onClose = () => console.log('Modal was requested to be closed.');
            return (
                <Modal
                    isOpen={boolean('Open', false)}
                    contentLabel={text(
                        'Label for content visible to screenreaders',
                        'My superb modal'
                    )}
                    onRequestClose={onClose}
                >
                    Some content
                </Modal>
            );
        },
        {
            info: {
                text: `
                ## Usage information

                This component is a wrapper around [react-modal](http://reactcommunity.org/react-modal/#usage).
                
                * You can pass other props according to their definition, apart from classes.
                * For accessibility reasons you need to __initialise__ modal use in your app. 
                For more info see [app element](http://reactcommunity.org/react-modal/accessibility/#app-element) in thier documentation. 
                To do this:

                > import OneUI from '@textkernel/oneui';

                > OneUI.__initModal__(appElementSelector);

                

                To see the modal in action here, change the Open prop to 'true' in the Knobs section.
                `
            }
        }
    );
