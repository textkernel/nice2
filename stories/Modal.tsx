import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { text, withKnobs } from '@storybook/addon-knobs';
import { Modal, Button } from '@textkernel/oneui';
import { StoreInjector } from '../src/packages/storybook/withStore';

storiesOf('Atoms|Modal', module)
    .addDecorator(withKnobs)
    .addParameters(
        StoreInjector.withStore({
            isOpen: false,
        })
    )
    .add(
        'Modal',
        (storyContext) => {
            const store = storyContext?.parameters.getStore();
            const onClose = () => {
                store.set({ isOpen: false });
                console.log('Modal was requested to be closed.');
            };
            return (
                <div>
                    <Button onClick={() => store.set({ isOpen: true })}>Open Modal</Button>
                    <Modal
                        isOpen={store.get('isOpen')}
                        contentLabel={text(
                            'Label for content visible to screenreaders',
                            'My superb modal'
                        )}
                        onRequestClose={onClose}
                    >
                        Some content
                    </Modal>
                </div>
            );
        },
        {
            info: {
                text: `
                ## Usage information

                This component is a wrapper around [react-modal](http://reactcommunity.org/react-modal/#usage).

                * You can pass other props according to their definition, apart from classes.
                * For accessibility reasons you need to __initialise__ modal use in your app.
                For more info see [app element](http://reactcommunity.org/react-modal/accessibility/#app-element) in thier documentation.
                To do this call the following function once:

                > __Modal.setAppElement__(appElementSelector);
                `,
            },
        }
    );
