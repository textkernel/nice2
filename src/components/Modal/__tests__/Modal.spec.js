import React from 'react';
import toJson from 'enzyme-to-json';
import Modal from '../Modal';

describe('Modal', () => {
    it('should render correctly', () => {
        const wrapper = shallow(
            <Modal isOpen contentLabel="Content label">
                Some children
            </Modal>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
