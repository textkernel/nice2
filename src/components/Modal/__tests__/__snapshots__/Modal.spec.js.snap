// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render correctly 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className={
    Object {
      "afterOpen": "Modal__content--entered",
      "base": "Modal__content",
      "beforeClose": "Modal__content--exited",
    }
  }
  closeTimeoutMS={300}
  contentLabel="Content label"
  isOpen={true}
  overlayClassName={
    Object {
      "afterOpen": "Modal__overlay--entered",
      "base": "Modal__overlay",
      "beforeClose": "Modal__overlay--exited",
    }
  }
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  Some children
</Modal>
`;
